test/scripts/run_test.py

```diff
@@ -47,47 +48,51 @@ def analyse_instrumentation_test_result(result):
     total_num = 0
-    fail_num = 0
+    success_num = 0
+    p = re.compile(".*?XML test result file generated at (.*?)\. Total tests (\d+).*?passed (\d+)")
     with open(result, 'r') as f:
         for line in f:
-            m_failure = re.match('Tests run:\s+(\d+).*?Failures:\s+(\d+)', line)
-            m_succeed = re.match('OK.*?(\d+) test', line)
-            if m_failure is not None:
-                total_num = int(m_failure.group(1))
-                fail_num = int(m_failure.group(2))
-            if m_succeed is not None:
-                total_num = int(m_succeed.group(1))
-    return total_num, fail_num
+            m = re.match(p, line)
+            if m is not None:
+                shutil.copyfile(m.group(1), os.path.join(os.path.split(result)[0], xml_file_name))
+                total_num = int(m.group(2))
+                success_num = int(m.group(3))
+    return total_num, success_num
 
 
-def exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file):
+def exec_cmd(cmd, device, target_package, result_file, logcat_file, test_path):
     adb = ['adb'] if device == None else ['adb', '-s', device]
     clean_logcat = ['logcat', '-c']
     subprocess.call(adb + clean_logcat)
     with open(result_file, 'a+') as rf:
-        subprocess.call(adb + am_cmd, stdout=rf)
+        subprocess.call(cmd, cwd=test_path, stdout=rf ,stderr=rf)
     logcat_cmd = ['logcat', '-d', target_package]
     with open(logcat_file, 'a+') as lf:
         subprocess.call(adb + logcat_cmd, stdout=lf)
 
 
-def run_cases(mode, module, target_package, log_dir, device, cases=None):
+def run_cases(mode, module, target_package, log_dir, device, test_path, cases=None):
     print '\n> running ' + mode.value + ' cases on device', device
     result_file = os.path.join(log_dir, module + '-' + mode.value + '-result-' + LOGCAT_SUFFIX)
     logcat_file = os.path.join(log_dir, module + '-' + mode.value + '-logcat-' + LOGCAT_SUFFIX)
 
     if mode == TestMode.INSTRUMENTATION:
+        adb = ['adb'] if device == None else ['adb', '-s', device]
         for case in cases:
             am_cmd = ['shell', 'am', 'instrument', '-w', '-r', '-e', 'debug', 'false', '-e',
                       'class', target_package + '.' + case,
                       target_package + '.test/android.test.InstrumentationTestRunner']
-            exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file)
+            exec_cmd(adb + am_cmd, device, target_package, result_file, logcat_file, test_path)
     elif mode == TestMode.UNIT:
-        am_cmd = ['shell', 'am', 'instrument', '-w', '-r', '-e', 'debug', 'false',
-                  target_package + '.test/android.support.test.runner.AndroidJUnitRunner']
-        exec_am_cmd(am_cmd, device, target_package, result_file, logcat_file)
+        if device is not None:
+            os.environ['ANDROID_SERIAL'] = device
+        cmd = [HOME_PATH + '/gradlew', '-q', 'clean']
+        subprocess.call(cmd, cwd=test_path)
```

Comment:
i believe there is a `exec_cmd` method to do this job.

Reply:
done